datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -80,92 +75,85 @@ struct ArrayHashSet {
     map: HashMap<usize, (), ()>,
 }
 
-struct ArraySet<T> {
-    array: T,
-    hash_set: ArrayHashSet,
-}
-
-impl<T> ArraySet<T>
-where
-    T: Array + From<ArrayData>,
-{
-    fn new(array: &T, hash_set: ArrayHashSet) -> Self {
-        Self {
-            array: downcast_array(array),
-            hash_set,
+impl ArrayHashSet {
+    /// Checks if values in `v` are contained in the `in_array` using this hash set for lookup.
+    fn contains(
+        &self,
+        v: &dyn Array,
+        in_array: &dyn Array,
+        negated: bool,
+    ) -> Result<BooleanArray> {
+        // Null type comparisons always return null (SQL three-valued logic)
+        if v.data_type() == &DataType::Null || in_array.data_type() == &DataType::Null {
+            return Ok(BooleanArray::from(vec![None; v.len()]));
         }
-    }
-}
 
-impl<T> Set for ArraySet<T>
-where
-    T: Array + 'static,
-    for<'a> &'a T: ArrayAccessor,
-    for<'a> <&'a T as ArrayAccessor>::Item: IsEqual,
-{
-    fn contains(&self, v: &dyn Array, negated: bool) -> Result<BooleanArray> {
         downcast_dictionary_array! {
             v => {
-                let values_contains = self.contains(v.values().as_ref(), negated)?;
+                let values_contains = self.contains(v.values().as_ref(), in_array, negated)?;
                 let result = take(&values_contains, v.keys(), None)?;
                 return Ok(downcast_array(result.as_ref()))
             }
             _ => {}
         }
 
-        let v = v.as_any().downcast_ref::<T>().unwrap();
-        let in_array = &self.array;
         let has_nulls = in_array.null_count() != 0;
 
-        Ok(ArrayIter::new(v)
-            .map(|v| {
-                v.and_then(|v| {
-                    let hash = v.hash_one(&self.hash_set.state);
-                    let contains = self
-                        .hash_set
-                        .map
-                        .raw_entry()
-                        .from_hash(hash, |idx| in_array.value(*idx).is_equal(&v))
-                        .is_some();
-
-                    match contains {
-                        true => Some(!negated),
-                        false if has_nulls => None,
-                        false => Some(negated),
-                    }
-                })
+        let mut hashes_buf = vec![0u64; v.len()];
```

Comment:
As a follow on PR we could potentially look into reusing this hashes_buf -- aka rather than reallocating each invocations of `contains` instead make a field (probably needs to be a `Mutex` or something) that is a Vec